plugins/geofence/geofence_impl.cpp

```diff
@@ -0,0 +1,219 @@
+#include "geofence_impl.h"
+#include "global_include.h"
+#include "log.h"
+
+namespace dronecode_sdk {
+
+GeofenceImpl::GeofenceImpl(System &system) :
+    PluginImplBase(system),
+    _mavlink_geofence_item_messages()
+{
+    _parent->register_plugin(this);
+}
+
+GeofenceImpl::~GeofenceImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+void GeofenceImpl::init()
+{
+    using namespace std::placeholders; // for `_1`
+
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_MISSION_REQUEST_INT,
+        std::bind(&GeofenceImpl::process_mission_request_int, this, _1),
+        (void *)this);
+
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_MISSION_ACK,
+        std::bind(&GeofenceImpl::process_mission_ack, this, _1),
+        (void *)this);
+}
+
+void GeofenceImpl::deinit()
+{
+    _parent->unregister_all_mavlink_message_handlers(this);
+}
+
+void GeofenceImpl::enable() {}
+
+void GeofenceImpl::disable() {}
+
+void GeofenceImpl::send_geofence_async(
+    const std::vector<std::shared_ptr<Geofence::Polygon>> &polygons,
+    const Geofence::result_callback_t &callback)
+{
+    if (_active) {
+        report_geofence_result(callback, Geofence::Result::BUSY);
+        return;
+    }
+
+    if (!_parent->does_support_mission_int()) {
+        LogDebug() << "Mission int messages not supported";
+        report_geofence_result(callback, Geofence::Result::ERROR);
+        return;
+    }
+
+    _active = true;
+
+    assemble_mavlink_messages(polygons);
+
+    LogDebug() << "size afterwards is: " << _mavlink_geofence_item_messages.size();
+
+    mavlink_message_t message;
+    mavlink_msg_mission_count_pack(_parent->get_own_system_id(),
+                                   _parent->get_own_component_id(),
+                                   &message,
+                                   _parent->get_system_id(),
+                                   _parent->get_autopilot_id(),
+                                   _mavlink_geofence_item_messages.size(),
+                                   MAV_MISSION_TYPE_FENCE);
+
+    LogDebug() << "About to send " << _mavlink_geofence_item_messages.size() << " geofence items";
+
+    if (!_parent->send_message(message)) {
+        report_geofence_result(callback, Geofence::Result::ERROR);
+    }
+
+    _result_callback = callback;
+
+    _parent->register_timeout_handler(
```

Comment:
@julianoes Also I wasn't really sure what the purpose of this timeout was - could you elaborate on its function?

Reply:
We need to implement some timeouts according to https://mavlink.io/en/services/mission.html#operations to make sure we retransmit messages in case they're lost.